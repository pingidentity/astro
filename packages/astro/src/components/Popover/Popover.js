// TODO: popover-deprecate Remove when popover is deprecated in 1.0.0.
// Replace with PopoverContainer
import React, { forwardRef } from 'react';
import Tippy from '@tippyjs/react/headless';
import PropTypes from 'prop-types';
import Box from '../Box';
import useDeprecationWarning from '../../hooks/useDeprecationWarning';

/**
 * **WARNING: Will be deprecated in Astro 1.0.0, use `PopoverMenu` instead.**
 *
 * Popover component used for popover button menus and help hints.
 * Built using Tippy.js [Tippy.js](https://atomiks.github.io/tippyjs/) and uses the
 * available [props from Tippy.js](https://atomiks.github.io/tippyjs/v6/all-props/).
*/

const Popover = forwardRef((props, ref) => {
  const {
    content,
    trigger,
    ...others
  } = props;

  useDeprecationWarning('`Popover` will be deprecated in Astro-UI 1.0.0, use `PopoverMenu` for menus instead. Or `Tooltip` when it is available.');

  return (
    <Tippy
      ref={ref}
      interactive
      render={attrs => (
        <Box {...attrs} variant="popover.container">
          <Box role="tooltip" variant="popover.content">
            {content}
          </Box>
          <Box
            className="tippy-arrow"
            data-popper-arrow=""
            variant="popover.arrow"
          />
        </Box>
      )}
      {...(trigger && { trigger })}
      {...others}
    />
  );
});

Popover.propTypes = {
  /** The content of the popover. */
  content: PropTypes.node,
  /** Placement of the popover. Note that TippyJs's flip modifier can change this
   *  to the opposite placement if it has more space. [More information on placement](https://atomiks.github.io/tippyjs/v6/all-props/#placement) */
  placement: PropTypes.string,
  /** Determines the events that cause the tooltip to show.
   * Multiple event names are separated by spaces.
   * [More information on trigger](https://atomiks.github.io/tippyjs/v6/all-props/#trigger) */
  trigger: PropTypes.string,
  /** Delay in ms once a trigger event is fired before a popover shows or hides.
   * [More information on delay](https://atomiks.github.io/tippyjs/v6/all-props/#delay) */
  delay: PropTypes.array,
};

Popover.defaultProps = {
  placement: 'top',
  trigger: 'mouseenter focus',
};

export default Popover;
