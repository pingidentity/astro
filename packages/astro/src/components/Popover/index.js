// TODO: popover-deprecate Remove when popover is deprecated in 1.0.0
export { default } from './Popover';
