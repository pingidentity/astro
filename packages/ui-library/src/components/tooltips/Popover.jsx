import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import _ from "underscore";

import popsOver from "../../util/behaviors/popsOver";

/**
 * @class Popover
 * @desc Popover creates a trigger component that controls the visibility of the content.
 *
 * @param {string} [data-id="popover"]
 *     To define the base "data-id" value for top-level HTML container.
 * @param {string} [className]
 *     CSS classes to set on the top-level HTML container.
 * @param {string} [triggerClassName]
 *     CSS classes to set on the link that triggers the popover.
 *
 * @param {object|string} [label]
 *     A string or JSX object that serves as the trigger label.
 *
 * @param {string} [placement]
 *     Placement using keywords top, left, right. (Default is bottom center)
 *
 * @param {boolean} [padded=false]
 *     If true, the content will be padded from the frame.
 *
 * @param {boolean} [open=false]
 *     If true, tooltip is open or else closed.
 * @param {function} [onToggle]
 *     Callback to be triggered when trigger is clicked.
 **/

class PopoverBase extends React.Component {
    static displayName = "Popover";

    static propTypes = {
        "data-id": PropTypes.string,
        children: PropTypes.node,
        className: PropTypes.string,
        label: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
        onKeyDown: PropTypes.func,
        onToggle: PropTypes.func,
        padded: PropTypes.bool,
        open: PropTypes.bool,
        placement: PropTypes.string,
        triggerClassName: PropTypes.string,
    };

    static defaultProps = {
        "data-id": "popover",
        label: "Link",
        onKeyDown: _.noop,
        onToggle: _.noop,
        placement: "",
        triggerClassName: "",
        padded: false,
    };

    /*
     * Check if placement contains a word (like top, right, or left)
     */
    _hasPlacement = word =>
        this.props.placement && this.props.placement.search(word) >= 0;

    /*
     * Create object for classnames to add modifiers to a block
     */
    _getModifiers = block => {
        const modifiers = {};
        modifiers[`${block}--top`] = this._hasPlacement("top");
        modifiers[`${block}--left`] = this._hasPlacement("left");
        modifiers[`${block}--right`] = this._hasPlacement("right");

        return modifiers;
    }

    renderContent = () => this.props.children;

    /*
     * Return of content based on props.open.
     *
     * @return {React.Component} the React component to be used as tooltip content
     */
    _content = () => {
        var frameClassName = classnames(
            "popup-frame popup-frame--pointer",
            this.props.className,
            this._getModifiers("popup-frame"), {
                "popup-frame--padded": this.props.padded,
            }
        );

        return this.props.open ? (
            <div className="popover__container">
                <div className={frameClassName}>{this.renderContent()}</div>
            </div>
        ) : null;
    };

    render() {
        var containerClassName = classnames(
            "popover",
            this.props.className,
            this._getModifiers("popover")
        );

        return (
            <div className={containerClassName} data-id={this.props["data-id"]} ref="content">
                <a
                    className={classnames(
                        "popover__trigger",
                        this.props.triggerClassName,
                        {
                            active: this.props.open
                        }
                    )}
                    data-id={`${this.props["data-id"]}-trigger`}
                    onClick={this.props.onToggle}
                    onKeyDown={this.props.onKeyDown}
                    title={this.props.title}
                >
                    {this.props.label}
                </a>
                {this._content()}
            </div>
        );
    }
}

const Popover = popsOver(PopoverBase);
Popover.Base = PopoverBase;

module.exports = Popover;
