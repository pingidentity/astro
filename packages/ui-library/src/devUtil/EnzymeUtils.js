import { mount } from "enzyme";


export const snapshotDataIds = node => {
    const elements = node.find("[data-id]").hostNodes(); // Find only actual HTML elements

    const ids = elements.map(element => element.prop("data-id")).filter(item => item !== "");
    expect(ids).toMatchSnapshot();
};

export const mountSnapshotDataIds = node => snapshotDataIds(mount(node));
