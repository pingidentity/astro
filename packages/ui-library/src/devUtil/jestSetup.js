
// This is needed for testing the modal and DetailsTooltip component.
global.getSelection = jest.fn();
global.getSelection.mockReturnValue({
    toString: () => "",
});
global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
        nodeName: "BODY",
        ownerDocument: document,
    },
});
