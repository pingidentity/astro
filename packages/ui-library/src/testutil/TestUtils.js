/*eslint-disable valid-jsdoc*/

"use strict";

/**
 * @module util/TestUtils
 * @desc A collection of ReactJS test utils, to complement the ones provided by the
 * framework. This is a copy of the TestUtils in the library, just because this class
 * is used in the tests and should work with the same version of React used in the WP
 * and not in the library.
 */

var ReactDOM = require("react-dom");
var React = require("react");
var ReactTestUtils = require("react-addons-test-utils");
var _ = require("underscore");
var assign = require("object-assign");

var TestUtils = {

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithDataId
     * @desc Return all DOM nodes children of the given tree with the specified dataId.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {array} - of DOM node objects which match the given criteria
     */
    scryRenderedDOMNodesWithDataId: function (parent, dataId) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.getAttribute("data-id") === dataId) {
            nodes = [parentNode];
        }
        else {
            var selectorValue = dataId
                .replace(/\[/g, "\\[")
                .replace(/\]/g, "\\]")
                .replace(/'/g, "\\'");
            nodes = parentNode.querySelectorAll("[data-id=" + selectorValue + "]");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithDataId
     * @desc Return the single DOM node child of the given tree with the specified dataId.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {object} - the DOM node object which matches the given criteria
     */
    findRenderedDOMNodeWithDataId: function (parent, dataId) {
        var nodes = this.scryRenderedDOMNodesWithDataId(parent, dataId);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with dataId: " +
                dataId + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithTag
     * @desc Return all DOM nodes children of the given tree with the specified tag name.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target elements
     * @return {array} - array of DOM nodes object which match the given criteria
     */
    scryRenderedDOMNodesWithTag: function (parent, tagName) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.tagName.toLowerCase() === tagName.toLowerCase()) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll(tagName);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithTag
     * @desc Return the single DOM node child of the given tree with the specified tag name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithTag: function (parent, tagName) {
        var nodes = this.scryRenderedDOMNodesWithTag(parent, tagName);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with tag name: " +
                tagName + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithName
     * @desc Return all DOM nodes children of the given tree with the specified name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target elements
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithName: function (parent, name) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.name === name) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll("[name='" + name.replace("'", "\\'") + "']");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithName
     * @desc Return the single DOM node child of the given tree with the specified name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithName: function (parent, name) {
        var nodes = this.scryRenderedDOMNodesWithName(parent, name);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with name: " +
                name + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithClass
     * @desc Return all DOM nodes children of the given tree with the specified CSS class name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target elements (or a concatenation of classes, separated by space)
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithClass: function (parent, className) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode = ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
            ? ReactDOM.findDOMNode(parent)
            : parent;

        var nodes;

        // first check if the root element has the provided classes;
        // if not, build a selector and look for any children

        var parentMatches = _.every(className.split(" "), function (name) {
            return (" " + parentNode.className + " ").indexOf(" " + name + " ") > -1;
        });
        if (parentMatches) {
            nodes = [parentNode];
        }
        else {
            var selector = _.reduce(className.split(" "), function (memo, name) {
                return memo + "." + name;
            }, "");
            nodes = parentNode.querySelectorAll(selector);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithClass
     * @desc Return the single DOM node child of the given tree with the specified CSS class name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithClass: function (parent, className) {
        var nodes = this.scryRenderedDOMNodesWithClass(parent, className);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with class name: " +
                className + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedComponentsWithType
     * @desc Finds all instances of components with type equal to `componentType`.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {array} an array of all the matches.
     */
    scryRenderedComponentsWithType: function (root, componentType) {
        return ReactTestUtils.scryRenderedComponentsWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.findRenderedComponentWithType
     * @desc Same as `scryRenderedComponentsWithType` but expects there to be one result
     * and returns that one result, or throws exception if there is any other
     * number of matches besides one.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {!ReactComponent} The one match.
     */
    findRenderedComponentWithType: function (root, componentType) {
        return ReactTestUtils.findRenderedComponentWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.captureGlobalListener
     * @desc Attaches proxy function for `window.addEvenListener` which will save reference to actual provided callback when
     * event type match filter. Useful for testing ReactJS component global (window) callbacks behaviour.
     *
     * @param {string} eventType event to save callback for. case insensitive.
     * @param {object} [obj=window] optional object to capture listener from. Like `document`. If not provided - window object
     * @returns {function} - when executed will trigger actual provided callback to `window.addEventListener`
     *
     * @example
     * //capture registered callback from reactjs control
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown");
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     *
     * @example
     * //capture registered callback from reactjs control on `document` object
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown", document);
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     */
    captureGlobalListener: function (eventType, obj) {

        var captured;

        var target = obj || window;

        target.addEventListener = function (event, listener) {
            if (event.toLowerCase() === eventType.toLowerCase()) {
                captured = listener;
            }
        };

        return function () {
            return captured.apply(this, arguments);
        };
    }

    /**
     * @desc Return all components in the supplied tree with the specified dataId.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {ReactComponent[]} - all components which match the given criteria
     */
    /*
     scryRenderedDOMComponentsWithDataId: function (tree, dataId) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instId = ReactDOM.findDOMNode(inst).getAttribute("data-id");
     return ReactTestUtils.isDOMComponent(inst) && (instId === dataId);
     });

     return components;
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified dataId.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithDataId: function (tree, dataId) {
     var components = TestUtils.scryRenderedDOMNodesWithDataId(tree, dataId);

     if (components.length > 1) {
     throw new Error("Found more than one component with dataId: " +
     dataId + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified name.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} name - the value of the name attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithName: function (tree, name) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instName = ReactDOM.findDOMNode(inst).name;
     return ReactTestUtils.isDOMComponent(inst) && (name === instName);
     });

     if (components.length > 1) {
     throw new Error("Found more than one component with name: " +
     name + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     }
     */
};

/**
 * @class UpdatePropsWrapper
 *
 * @desc
 * A simple wrapper to be used for unit testing, when you need to update
 * a property on a rendered component. Starting with React 0.14.x,
 * directly setting or replacing a prop on a rendered component is deprecated.
 * Instead, use this wrapper and call _setProps(newProps) on it
 * to have the wrapped component re-render with the new property(ies).
 * The ref attribute on the wrapper is set to "wrapper".
 *
 * @Example
 *  // initial render
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={If} test={true} />
 *  );
 *  // update the property through re-render
 *  component._setProps({test: false});
 *
 * @Example
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={TabbedSections}
 *              onSectionChange={jest.genMockFunction()}
 *              selectedIndex={-1}>
 *          <div data-id="section1" title="section 1">section 1</div>
 *          <div data-id="section2" title="section 2">section 2</div>
 *      </UpdatePropsWrapper>
 *  );
 *  component._setProps({ selectedIndex: 0 });
 *
 */
TestUtils.UpdatePropsWrapper = React.createClass({
    /**
     * Re-render the wrapped component using the new property.
     * Use this function to avoid setting the props directly on the component (anti pattern).
     * @param {object} props - new property(ies) to set
     */
    _setProps: function (props) {
        this.setState(props);
    },
    getInitialState: function () {
        return assign({ ref: "wrapper" }, this.props);
    },
    render: function () {
        return React.createElement(
            this.props.type, // this is the type of element to create, provided as prop
            this.state // this is the rest of props to set on the element
        );
    }
});

module.exports = TestUtils;
