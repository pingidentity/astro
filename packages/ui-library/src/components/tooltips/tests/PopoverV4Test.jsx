import { handleOpen } from "../../../util/behaviors/popsOver.js";

import React from "react";
import ReactDOM from "react-dom";
import ReactTestUtils from "react-dom/test-utils";
import { mountSnapshotDataIds } from "../../../devUtil/EnzymeUtils";
import TestUtils from "../../../testutil/TestUtils";
import KeyboardUtils from "../../../util/KeyboardUtils";
import Popover from "../Popover";
import _ from "underscore";
const Wrapper = TestUtils.UpdatePropsWrapper;

window.__DEV__ = true;

jest.dontMock("../Popover");
jest.dontMock("../../../util/EventUtils.js");
jest.dontMock("../../../util/Utils.js");

jest.mock("popper.js");
jest.mock("react-portal");

describe("Popover", function() {
    window.addEventListener = jest.fn();
    window.removeEventListener = jest.fn();
    beforeEach(function() {
        window.addEventListener.mockClear();
        window.removeEventListener.mockClear();
    });

    it("data-id's don't change", () => {
        mountSnapshotDataIds(
            <Popover
                label="BORK"
                open
            >
                Just some stuff
            </Popover>
        );
    });

    it("renders open state", function() {
        const component = ReactTestUtils.renderIntoDocument(
            <div>
                <Popover label="hello" open={true}>
                    <div className="content">Popover</div>
                </Popover>
            </div>
        );
        const popoverContainer = TestUtils.scryRenderedDOMNodesWithClass(
            component,
            "popover-display"
        );
        const popupFrame = TestUtils.scryRenderedDOMNodesWithClass(
            component,
            "popup-frame"
        );
        const content = TestUtils.scryRenderedDOMNodesWithClass(
            component,
            "content"
        );

        expect(popoverContainer.length).toBe(1);
        expect(popupFrame.length).toBe(1);
        expect(content.length).toBe(1);
    });

    it("renders closed state", function() {
        const component = ReactTestUtils.renderIntoDocument(
            <div>
                <Popover label="hello" open={false}>
                    <div className="content">Popover</div>
                </Popover>
            </div>
        );
        const popoverTarget = TestUtils.scryRenderedDOMNodesWithClass(
            component,
            "popover__trigger"
        );

        expect(popoverTarget.length).toBe(1);

        const popoverContainer = TestUtils.scryRenderedDOMNodesWithClass(
            component,
            "popover__container"
        );
        const popupFrame = TestUtils.scryRenderedDOMNodesWithClass(
            component,
            "popup-frame"
        );
        const content = TestUtils.scryRenderedDOMNodesWithClass(
            component,
            "content"
        );

        expect(popoverContainer.length).toBe(0);
        expect(popupFrame.length).toBe(0);
        expect(content.length).toBe(0);
    });

    it("renders with top and left classes", function() {
        const component = ReactTestUtils.renderIntoDocument(
            <div>
                <Popover label="hello" open={true} placement="top left">
                    <div className="content">Popover</div>
                </Popover>
            </div>
        );
        const withLeftClass = TestUtils.scryRenderedDOMNodesWithClass(
            component,
            "popover--left"
        );
        const withTopClass = TestUtils.scryRenderedDOMNodesWithClass(
            component,
            "popover--top"
        );

        expect(withLeftClass.length).toBe(1);
        expect(withTopClass.length).toBe(1);
    });

    it("renders with right class", function() {
        const component = ReactTestUtils.renderIntoDocument(
            <div>
                <Popover label="hello" open={true} placement="right">
                    <div className="content">Popover</div>
                </Popover>
            </div>
        );
        const withRightClass = TestUtils.scryRenderedDOMNodesWithClass(
            component,
            "popover--right"
        );

        expect(withRightClass.length).toBe(1);
    });

    it("renders with top and left placement using a portal", function() {
        const component = ReactTestUtils.renderIntoDocument(
            <div>
                <Popover label="hello" open={true} placement="top left">
                    <div className="content">Popover</div>
                </Popover>
            </div>
        );
        const withLeftClass = TestUtils.scryRenderedDOMNodesWithClass(
            component,
            "popover--left"
        );
        const withTopClass = TestUtils.scryRenderedDOMNodesWithClass(
            component,
            "popover--top"
        );

        expect(withLeftClass.length).toBe(1);
        expect(withTopClass.length).toBe(1);
    });

    it("renders with right placement using a portal", function() {
        const component = ReactTestUtils.renderIntoDocument(
            <div>
                <Popover label="hello" open={true} placement="right">
                    <div className="content">Popover</div>
                </Popover>
            </div>
        );
        const withRightClass = TestUtils.scryRenderedDOMNodesWithClass(
            component,
            "popover--right"
        );

        expect(withRightClass.length).toBe(1);
    });

    it("renders with center placement using a portal", function() {
        const component = ReactTestUtils.renderIntoDocument(
            <div>
                <Popover label="hello" open={true} placement="center">
                    <div className="content">Popover</div>
                </Popover>
            </div>
        );
        const withRightClass = TestUtils.scryRenderedDOMNodesWithClass(
            component,
            "popover--right"
        );

        expect(withRightClass.length).toBe(0);
    });

    it("notifies toggle when clicking trigger", function() {
        const callback = jest.fn();

        const component = ReactTestUtils.renderIntoDocument(
            <div>
                <Popover onToggle={callback} label="hello" open={false}>
                    <div className="content">Popover</div>
                </Popover>
            </div>
        );

        const target = TestUtils.findRenderedDOMNodeWithClass(
            component,
            "popover__trigger"
        );

        expect(callback).not.toBeCalled(); //make sure callback was NOT triggered

        ReactTestUtils.Simulate.click(target, {});

        expect(callback).toBeCalled(); //make sure callback was triggered
    });

    it("registers global listener on mount if component is open", function() {
        //let's override defer or execute func immediately for tests
        _.defer = function(func) {
            func();
        };

        ReactTestUtils.renderIntoDocument(
            <div>
                <Popover label="hello" open={true}>
                    <div className="content">Popover</div>
                </Popover>
            </div>
        );

        expect(
            TestUtils.mockCallsContains(window.addEventListener, "click")
        ).toBe(true);
        expect(
            TestUtils.mockCallsContains(window.addEventListener, "keydown")
        ).toBe(true);
    });

    it("does not register global listener on mount if component is closed", function() {
        //let's override defer or execute func immediately for tests
        _.defer = function(func) {
            func();
        };

        ReactTestUtils.renderIntoDocument(
            <div>
                <Popover label="hello" open={false}>
                    <div className="content">Popover</div>
                </Popover>
            </div>
        );

        expect(
            TestUtils.mockCallsContains(window.addEventListener, "click")
        ).toBe(false);
        expect(
            TestUtils.mockCallsContains(window.addEventListener, "keydown")
        ).toBe(false);
    });

    it("unregister global listeners on unmount", function() {
        const component = ReactTestUtils.renderIntoDocument(
            <div>
                <Popover label="hello" open={true}>
                    <div className="content">Popover</div>
                </Popover>
            </div>
        );

        expect(
            TestUtils.mockCallsContains(window.removeEventListener, "click")
        ).toBe(false);
        expect(
            TestUtils.mockCallsContains(window.removeEventListener, "keydown")
        ).toBe(false);

        //trigger unmount
        ReactDOM.unmountComponentAtNode(
            ReactDOM.findDOMNode(component).parentNode
        );

        expect(
            TestUtils.mockCallsContains(window.removeEventListener, "click")
        ).toBe(true);
        expect(
            TestUtils.mockCallsContains(window.removeEventListener, "keydown")
        ).toBe(true);
    });

    it("unregisters listener when transitioning from open to closed", function() {
        const component = ReactTestUtils.renderIntoDocument(
            <Wrapper type={Popover} label="hello" open={true}>
                <div className="content">Popover</div>
            </Wrapper>
        );

        component._setProps({ open: false });

        expect(
            TestUtils.mockCallsContains(window.removeEventListener, "click")
        ).toBe(true);
        expect(
            TestUtils.mockCallsContains(window.removeEventListener, "keydown")
        ).toBe(true);
    });

    it("registers listener when transitioning from closed to open", function() {
        const component = ReactTestUtils.renderIntoDocument(
            <Wrapper type={Popover} label="hello" open={false}>
                <div className="content">Popover</div>
            </Wrapper>
        );

        component._setProps({ open: true });

        expect(
            TestUtils.mockCallsContains(window.addEventListener, "click")
        ).toBe(true);
        expect(
            TestUtils.mockCallsContains(window.addEventListener, "keydown")
        ).toBe(true);
    });

    it("registers and unregisters listener when transitioning -- stateful", function() {
        const component = ReactTestUtils.renderIntoDocument(
            <Wrapper type={Popover} label="hello">
                <div className="content">Popover</div>
            </Wrapper>
        );

        const trigger = TestUtils.findRenderedDOMNodeWithClass(
            component,
            "popover__trigger"
        );
        ReactTestUtils.Simulate.click(trigger, {});
        expect(
            TestUtils.mockCallsContains(window.addEventListener, "click")
        ).toBe(true);
        expect(
            TestUtils.mockCallsContains(window.addEventListener, "keydown")
        ).toBe(true);

        ReactTestUtils.Simulate.click(trigger, {});
        expect(
            TestUtils.mockCallsContains(window.removeEventListener, "click")
        ).toBe(true);
        expect(
            TestUtils.mockCallsContains(window.removeEventListener, "keydown")
        ).toBe(true);
    });

    it("triggers callback when clicked outside", function() {
        const callback = jest.fn();
        global.getSelection = jest.fn();
        global.getSelection.mockReturnValue({
            toString: () => "",
        });

        ReactTestUtils.renderIntoDocument(
            <div>
                <Popover onToggle={callback} label="hello" open={true}>
                    <div className="content">Popover</div>
                </Popover>
            </div>
        );

        const handler = TestUtils.findMockCall(
            window.addEventListener,
            "click"
        )[1];
        const e = {
            target: { parentNode: document.body },
            stopPropagation: jest.fn(),
            preventDefault: jest.fn()
        };

        expect(callback).not.toBeCalled();

        //click outside
        handler(e);

        expect(callback).toBeCalled();
    });

    it("triggers callback when global key event fires", function() {
        const callback = jest.fn();

        ReactTestUtils.renderIntoDocument(
            <div>
                <Popover onToggle={callback} label="hello" open={true}>
                    <div className="content">Popover</div>
                </Popover>
            </div>
        );

        const handler = TestUtils.findMockCall(
            window.addEventListener,
            "keydown"
        )[1];
        const e = {
            target: { parentNode: document.body },
            stopPropagation: jest.fn(),
            preventDefault: jest.fn(),
            keyCode: KeyboardUtils.KeyCodes.ESC
        };
        const eWrong = {
            target: { parentNode: document.body },
            stopPropagation: jest.fn(),
            preventDefault: jest.fn(),
            keyCode: 2
        };

        expect(callback).not.toBeCalled();
        handler(eWrong);
        expect(callback).not.toBeCalled();

        //click outside
        handler(e);

        expect(callback).toBeCalled();
    });

    it("call handleOpen", function() {
        const callback = jest.fn();

        expect(callback).not.toBeCalled();
        handleOpen({ onOpen: callback });
        expect(callback).toBeCalled();
    });
});
