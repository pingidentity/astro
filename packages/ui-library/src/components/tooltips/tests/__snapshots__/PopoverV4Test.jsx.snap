// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover data-id's don't change 1`] = `
Array [
  "popover",
  "popover-trigger",
  "popup-frame",
]
`;
